### PR TITLE
Feature/d3 asim 1758

### PR DESCRIPTION
**Pull request by hannesdiedrich**

### CODE REVIEW

File: src/d3a/models/market/two_sided_pay_as_bid.py [hunk 1]
```diff
@@ -120,70 +171,44 @@ def accept_bid(self, bid: Bid, energy: float = None,
...
             self._update_stats_after_trade(trade, bid, bid.buyer, already_tracked)
             log.info(f"[TRADE][BID] [{self.time_slot_str}] {trade}")
-            final_bid = bid._replace(price=final_price)
-            trade = trade._replace(offer=final_bid)
 
         self._notify_listeners(MarketEvent.BID_TRADED, bid_trade=trade)
-        if not trade.residual:
-            self._notify_listeners(MarketEvent.BID_DELETED, bid=market_bid)
+        # if not trade.residual:
+        #     self._notify_listeners(MarketEvent.BID_DELETED, bid=market_bid)
```
Comment (faizan2590) [inline src/d3a/models/market/two_sided_pay_as_bid.py]: Please remove before merging.

File: src/d3a/models/strategy/area_agents/two_sided_pay_as_bid_engine.py [hunk 2]
```diff
@@ -45,17 +47,17 @@ def _forward_bid(self, bid):
...
+            energy=bid.energy,
+            buyer=self.owner.name,
+            seller=self.markets.target.name,
             original_bid_price=bid.original_bid_price,
             buyer_origin=bid.buyer_origin
         )
-        bid_coupling = BidInfo(bid, forwarded_bid)
-        self.forwarded_bids[forwarded_bid.id] = bid_coupling
-        self.forwarded_bids[bid.id] = bid_coupling
+        # assert False
```
Comment (faizan2590) [inline src/d3a/models/strategy/area_agents/two_sided_pay_as_bid_engine.py]: Please remove before merging. Thanks!

File: src/d3a/models/market/two_sided_pay_as_bid.py [hunk 3]
```diff
@@ -120,70 +171,44 @@ def accept_bid(self, bid: Bid, energy: float = None,
...
+        trade_price = self.determine_bid_price(trade_offer_info, energy)
+        bid = bid._replace(price=trade_price)
 
         # Do not adapt grid fees when creating the bid_trade_info structure, to mimic
         # the behavior of the forwarded bids which use the source market fee.
         updated_bid_trade_info = GridFees.propagate_original_offer_info_on_bid_trade(
                           trade_offer_info, 0.0)
 
+        # Delete the accepted bid from self.bids:
+        self.bids.pop(bid.id, None)
```
Comment (faizan2590) [inline src/d3a/models/market/two_sided_pay_as_bid.py]:
```suggestion
        try:
            self.bids.pop(bid.id)
        except KeyError:
            raise BidNotFoundException
```

Comment (hannesdiedrich) [reply src/d3a/models/market/two_sided_pay_as_bid.py]: Actually, this is not needed at all, as it is done at the first line in accept_bid. Removing it. Thanks fort pointing it out!

Comment (hannesdiedrich) [reply src/d3a/models/market/two_sided_pay_as_bid.py]: I was wrong. This is again the accepted bid. I will incorporate your proposal.

File: tests/test_inter_area_agent.py [hunk 4]
```diff
@@ -97,9 +96,11 @@ def accept_offer(self, offer_or_id, buyer, *, energy=None, time=None, already_tr
             residual = Offer('res', offer.price, residual_energy, offer.seller,
                              seller_origin='res')
             traded = Offer(offer.id, offer.price, energy, offer.seller, seller_origin='res')
+            self.bids.pop(traded.id, None)
```
Comment (spyrostz) [inline tests/test_inter_area_agent.py]: Why is this needed here ?

File: src/d3a/models/market/two_sided_pay_as_bid.py [hunk 5]
```diff
@@ -120,70 +171,45 @@ def accept_bid(self, bid: Bid, energy: float = None,
...
-                      buyer_origin=bid.buyer_origin)
+            # partial bid trade
+            accepted_bid, residual_bid = self.split_bid(market_bid, energy, orig_price)
+            bid = accepted_bid
+
+            # Delete the accepted bid from self.bids:
+            try:
+                self.bids.pop(accepted_bid.id)
+            except KeyError:
+                raise BidNotFound
```
Comment (spyrostz) [inline src/d3a/models/market/two_sided_pay_as_bid.py]: You need to instantiate the exception class here, I am not sure that it works just with the class name (never tried it, but I have doubts that it works that way). Also an informative message does not hurt.

File: tests/test_market.py [hunk 6]
```diff
@@ -300,7 +300,6 @@ def test_balancing_market_negative_offer_trade(market=BalancingMarket(time_slot=
 
 
 def test_market_bid_trade(market=TwoSidedPayAsBid(time_slot=DateTime.now())):
-    # market = TwoSidedPayAsBid(time_slot=DateTime.now())
```
Comment (spyrostz) [inline tests/test_market.py]: You can remove this line

File: tests/test_inter_area_agent.py (hunk 4 above)
Comment (hannesdiedrich) [reply tests/test_inter_area_agent.py]: Its not, remnant of debugging. Will remove it, also in the accept_bid method. Thanks!

File: src/d3a/models/market/two_sided_pay_as_bid.py (hunk 5 above)
Comment (hannesdiedrich) [reply src/d3a/models/market/two_sided_pay_as_bid.py]: Will do, thanks!

File: src/d3a/models/strategy/__init__.py [hunk 7]
```diff
@@ -430,12 +430,13 @@ def event_bid_deleted(self, *, market_id, bid):
...
 
-    def event_bid_changed(self, *, market_id, existing_bid, new_bid):
+    def event_bid_split(self, *, market_id, original_bid, accepted_bid, residual_bid):
         assert ConstSettings.IAASettings.MARKET_TYPE is not 1, \
             "Invalid state, cannot receive a bid if single sided market is globally configured."
-        if new_bid.buyer != self.owner.name:
+        if accepted_bid.buyer != self.owner.name:
             return
-        self.add_bid_to_posted(market_id, bid=new_bid)
+        self.add_bid_to_posted(market_id, bid=accepted_bid)
```
Comment (spyrostz) [inline src/d3a/models/strategy/__init__.py]: Clarification comment: the accepted bid needs to be added to posted, in order to be accepted in the trade event afterwards, right ? Thanks in advance

File: src/d3a/models/strategy/area_agents/two_sided_pay_as_bid_engine.py [hunk 8]
```diff
@@ -153,34 +140,59 @@ def event_bid_deleted(self, *, bid):
...
+
+            local_split_bid, local_residual_bid = \
+                self.markets.source.split_bid(local_bid, accepted_bid.energy, original_bid_price)
+
+            #  add the new bids to forwarded_bids
+            self._add_to_forward_bids(local_residual_bid, residual_bid)
+            self._add_to_forward_bids(local_split_bid, accepted_bid)
+
+        elif market == self.markets.source and accepted_bid.id in self.forwarded_bids:
+            # bid in the source market was split, also split in the target market
```
Comment (spyrostz) [inline src/d3a/models/strategy/area_agents/two_sided_pay_as_bid_engine.py]: Better to add one more comment here, that this is the case that handles forwarded bids to markets other than the one that the trade will take place.

File: tests/test_market.py [hunk 9]
```diff
@@ -414,32 +413,14 @@ def test_market_trade_bid_partial(market=TwoSidedPayAsBid(time_slot=DateTime.now
...
-        'market_id': repr(market.id),
-        'bid_trade': repr(trade),
-    }
-    assert called.calls[1][1] == {
-        'market_id': repr(market.id),
-        'bid': repr(bid),
-    }
-
-
 def test_market_accept_bid_does_not_emit_bid_deleted_on_partial_bid(
```
Comment (spyrostz) [inline tests/test_market.py]: Maybe rename this to `test_market_accept_bid_emits_bid_split_on_partial_bid`

File: src/d3a/models/strategy/__init__.py (hunk 7 above)
Comment (hannesdiedrich) [reply src/d3a/models/strategy/__init__.py]: This is probably not needed, but you skared me when you said: "What happens, if the trade wasnt successfull" So i added the accepted bid to the posted bids, which gets deleted once traded. What do you say? Shall i remove this feature?

Comment (spyrostz) [reply src/d3a/models/strategy/__init__.py]: Doesn't hurt to have here IMO. By recalling the code flow in the trade event, this offer will move from posted to traded dict, therefore it has to be there. Thanks for clarifying
